### PR TITLE
Optimize VectorX.Clamp according to proposed change

### LINKED ISSUE #34017: Improve the codegen for Vector4.Clamp
https://github.com/SixLabors/ImageSharp makes use of the `System.Numerics.Vector*` and related types.

However, in https://github.com/SixLabors/ImageSharp/pull/1152 they are needing to replace the usage of `Vector4.Clamp` with a custom implementation due to the existing implementation being ~5x slower.

We should improve the codegen of this method in particular and also look at many of the other available methods and see where they could be trivially improved and/or accelerated.

### PR DESCRIPTION
**Pull request by Rekkonnect**
Closes #34017

### CODE REVIEW

**Comment by EgorBo** [comment]
Are `Vector4.Max\Min` actually vectorizes?
Your impl's codegen:
```asm
; Method Clamp1
G_M60623_IG01:
       sub      rsp, 56
       vzeroupper 
       vmovaps  qword ptr [rsp+20H], xmm6
       vmovaps  qword ptr [rsp+10H], xmm7
G_M60623_IG02:
       vmovss   xmm0, dword ptr [rdx]
       vmovss   xmm1, dword ptr [rdx+4]
       vmovss   xmm2, dword ptr [rdx+8]
       vmovss   xmm3, dword ptr [rdx+12]
       vmovss   xmm4, dword ptr [r8]
       vmovss   xmm5, dword ptr [r8+4]
       vmovss   xmm6, dword ptr [r8+8]
       vmovss   xmm7, dword ptr [r8+12]
       vucomiss xmm0, xmm4
       ja       SHORT G_M60623_IG04
G_M60623_IG03:
       jmp      SHORT G_M60623_IG05
G_M60623_IG04:
       vmovaps  xmm4, xmm0
G_M60623_IG05:
       vucomiss xmm1, xmm5
       ja       SHORT G_M60623_IG07
G_M60623_IG06:
       jmp      SHORT G_M60623_IG08
G_M60623_IG07:
       vmovaps  xmm5, xmm1
G_M60623_IG08:
       vucomiss xmm2, xmm6
       ja       SHORT G_M60623_IG10
G_M60623_IG09:
       jmp      SHORT G_M60623_IG11
G_M60623_IG10:
       vmovaps  xmm6, xmm2
G_M60623_IG11:
       vucomiss xmm3, xmm7
       ja       SHORT G_M60623_IG13
G_M60623_IG12:
       jmp      SHORT G_M60623_IG14
G_M60623_IG13:
       vmovaps  xmm7, xmm3
G_M60623_IG14:
       lea      rax, bword ptr [rsp]
       vmovss   dword ptr [rax], xmm4
       vmovss   dword ptr [rax+4], xmm5
       vmovss   dword ptr [rax+8], xmm6
       vmovss   dword ptr [rax+12], xmm7
       lea      rax, bword ptr [rsp]
       vmovss   xmm0, dword ptr [rax]
       vmovss   xmm1, dword ptr [rax+4]
       vmovss   xmm2, dword ptr [rax+8]
       vmovss   xmm3, dword ptr [rax+12]
       vmovss   xmm4, dword ptr [r9]
       vmovss   xmm5, dword ptr [r9+4]
       vmovss   xmm6, dword ptr [r9+8]
       vmovss   xmm7, dword ptr [r9+12]
       vucomiss xmm4, xmm0
       ja       SHORT G_M60623_IG16
G_M60623_IG15:
       jmp      SHORT G_M60623_IG17
G_M60623_IG16:
       vmovaps  xmm4, xmm0
G_M60623_IG17:
       vucomiss xmm5, xmm1
       ja       SHORT G_M60623_IG19
G_M60623_IG18:
       jmp      SHORT G_M60623_IG20
G_M60623_IG19:
       vmovaps  xmm5, xmm1
G_M60623_IG20:
       vucomiss xmm6, xmm2
       ja       SHORT G_M60623_IG22
G_M60623_IG21:
       jmp      SHORT G_M60623_IG23
G_M60623_IG22:
       vmovaps  xmm6, xmm2
G_M60623_IG23:
       vucomiss xmm7, xmm3
       ja       SHORT G_M60623_IG25
G_M60623_IG24:
       jmp      SHORT G_M60623_IG26
G_M60623_IG25:
       vmovaps  xmm7, xmm3
G_M60623_IG26:
       vmovss   dword ptr [rcx], xmm4
       vmovss   dword ptr [rcx+4], xmm5
       vmovss   dword ptr [rcx+8], xmm6
       vmovss   dword ptr [rcx+12], xmm7
       mov      rax, rcx
G_M60623_IG27:
       vmovaps  xmm6, qword ptr [rsp+20H]
       vmovaps  xmm7, qword ptr [rsp+10H]
       add      rsp, 56
       ret      
; Total bytes of code: 265
```
At the same time:
```csharp
public static Vector128<float> Clamp2(Vector128<float> value1,
    Vector128<float> min, Vector128<float> max)
{
    return Sse.Min(Sse.Max(value1, min), max);
}
```
Emits:
```asm
; Method Clamp2
G_M5676_IG01:
       vzeroupper 
G_M5676_IG02:
       vmovupd  xmm0, xmmword ptr [rdx]
       vmaxps   xmm0, xmm0, xmmword ptr [r8]
       vminps   xmm0, xmm0, xmmword ptr [r9]
       vmovupd  xmmword ptr [rcx], xmm0
       mov      rax, rcx
G_M5676_IG03:
       ret      
; Total bytes of code: 25
```

**Comment by AlFasGD** [comment]
@tannergooding would need to take a look at this, I'm not familiar with any method to further improve the codegen, I would assume that the proposed version would be just fine for this purpose, generating the best possible code.

**Comment by EgorBo** [comment]
Looks like none of the `[Intrinsics]` APIs in `Vector4_intrinsics.cs` actually accelerated or recognized by jit's importer.

**Comment by EgorBo** [comment]
Looks like `Vector4` vectorization is broken.

**Comment by JimBobSquarePants** [comment]
I'm seeing this on [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AcgK5QCmATgJYDGEA3OunAMy4LYDC2A3utgNxs4AFmwBZABQBKXv0EBfeQOVDcSAGzYAamRrAA9hTGcANmCgAHJJN36jYgB4AaHXsPHsUKgDtXdjzEoMEdpVUEI7ABeAD43e2N8cV9bdwckkMkXL19pV2DQxjQlNHQgA=) with the default settings.

``` asm
C.Clamp1(System.Numerics.Vector4, System.Numerics.Vector4, System.Numerics.Vector4)
    L0000: vzeroupper
    L0003: vmovupd xmm0, [esp+0x24]
    L0009: vmovupd xmm1, [esp+0x14]
    L000f: vmaxps xmm0, xmm0, xmm1
    L0013: vmovupd xmm1, [esp+0x4]
    L0019: vminps xmm0, xmm0, xmm1
    L001d: vmovupd [ecx], xmm0
    L0021: ret 0x30
```

We also saw a 5x improvement in ImageSharp when benchmarking the changes that led to the issue being raised.

https://github.com/SixLabors/ImageSharp/pull/1152/files#diff-c0645b9893e87cbacc35363b366826b1R55-R60

**Comment by EgorBo** [comment]
> I'm seeing this on [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AcgK5QCmATgJYDGEA3OunAMy4LYDC2A3utgNxs4AFmwBZABQBKXv0EBfeQOVDcSAGzYAamRrAA9hTGcANmCgAHJJN36jYgB4AaHXsPHsUKgDtXdjzEoMEdpVUEI7ABeAD43e2N8cV9bdwckkMkXL19pV2DQxjQlNHQgA=) with the default settings.

sharplab uses some stable .net core 3.0 runtime and I suspect something is broken in master (but maybe I am wrong)

**Comment by EgorBo** [comment]
Ah, I was using an old version of `System.Numerics.Vectors.dll` where `Vector4` type is not marked with `[Intrinsic]` so the master-runtime didn't recognize it as a SIMD.
Now the codegen is expected:
```asm
G_M52350_IG01:
       C5F877               vzeroupper
                                                ;; bbWeight=1    PerfScore 1.00
G_M52350_IG02:
       C5F91002             vmovupd  xmm0, xmmword ptr [rdx]
       C5F851C0             vsqrtps  xmm0, xmm0
       C5F91101             vmovupd  xmmword ptr [rcx], xmm0
       488BC1               mov      rax, rcx
                                                ;; bbWeight=1    PerfScore 18.25
G_M52350_IG03:
       C3                   ret
                                                ;; bbWeight=1    PerfScore 1.00

; Total bytes of code 19
```

**Comment by JimBobSquarePants** [comment]
Great news!

**Comment by EgorBo** [comment]
Also, probably makes sense to optimize `Lerp` (e.g. using fma)

**Comment by tannergooding** [comment]
@EgorBo, there are a lot of optimizations to still be had. As I mentioned in #34017, I'm happy to take codegen improvements for any of the System.Numerics methods (provided the appropriate tests still pass and there are benchmarks).
The DirectX Math library has a lot of methods we don't expose and a lot of optimizations (for both ARM and x86): https://github.com/microsoft/DirectXMath/blob/master/Inc/DirectXMathVector.inl#L5586

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs [hunk 1]
```diff
@@ -195,25 +195,8 @@ public static Vector4 Normalize(Vector4 vector)
...
-            float z = value1.Z;
-            z = (min.Z > z) ? min.Z : z;  // max(z, minz)
-            z = (max.Z < z) ? max.Z : z;  // min(z, maxz)
-
-            float w = value1.W;
-            w = (min.W > w) ? min.W : w;  // max(w, minw)
-            w = (max.W < w) ? max.W : w;  // min(w, minw)
-
-            return new Vector4(x, y, z, w);
+            // https://github.com/dotnet/runtime/issues/34017
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:198]: We shouldn't need to document the issue that was used for the improvement.
However, we should likely keep the `// We must follow HLSL behavior in the case user specified min value is bigger than max value.` comment

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs [hunk 2]
```diff
@@ -195,25 +195,8 @@ public static Vector4 Normalize(Vector4 vector)
...
+            return Vector4.Min(Vector4.Max(value1, min), max);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:199]: Would you be willing to make the same change to `Vector2` and `Vector3` as part of this PR (just so we can cover all 3 at once)?

**Comment by tannergooding** [comment]
Ideally we would also add some perf tests to https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/libraries/System.Numerics.Vectors/Perf_Vector4.cs (and Vector2/3 if updating as well) and provide some minimal before/after numbers to help track the perf improvement.

Would you be willing to do that as part of this?

If so, https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md contains the workflow for testing a benchmark against local dotnet/runtime changes. @adamsitnik is also a good resource for any questions that might come up.

**Comment by tannergooding** [comment]
CC. @pgovind for review

**Comment by tannergooding** [comment]
@AlFasGD, I think this is just missing some basic perf numbers from before/after the change so we can show it is a win.
It's needed largely just for tracking purposes to show we did our due diligence before merging.

**Comment by AlFasGD** [comment]
I don't know how to do that and the whole process is confusing me a bit, so I would hope somebody else could handle that.

**Comment by tannergooding** [comment]
I'll collect some basic numbers and share them back and directions on how to collect them in the future.

**Comment by AlFasGD** [comment]
Alright thanks for that

**Comment by tannergooding** [comment]
Numbers look good:
### Before: 
```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.778 (1909/November2018Update/19H2)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=5.0.100-preview.4.20222.23
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.22107, CoreFX 5.0.20.22107), X64 RyuJIT
  Job-BACAGA : .NET Core 5.0.0 (CoreCLR 5.0.20.22107, CoreFX 5.0.20.22107), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Runtime=.NET Core 5.0  Arguments=/p:DebugType=portable
Toolchain=netcoreapp5.0  IterationTime=250.0000 ms  MaxIterationCount=20
MinIterationCount=15  WarmupCount=1

|                 Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| Vector2.ClampBenchmark | 1.641 ns | 0.0316 ns | 0.0296 ns | 1.622 ns | 1.613 ns | 1.697 ns |     - |     - |     - |         - |
| Vector3.ClampBenchmark | 2.958 ns | 0.0235 ns | 0.0208 ns | 2.952 ns | 2.923 ns | 3.009 ns |     - |     - |     - |         - |
| Vector4.ClampBenchmark | 4.056 ns | 0.0742 ns | 0.0579 ns | 4.047 ns | 3.969 ns | 4.161 ns |     - |     - |     - |         - |
```

### After
```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.18363.778 (1909/November2018Update/19H2)
AMD Ryzen 9 3900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=5.0.100-preview.2.20176.6
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.22107, CoreFX 5.0.20.22107), X64 RyuJIT
  Job-JIKNIA : .NET Core 5.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:DebugType=portable  Toolchain=CoreRun
IterationTime=250.0000 ms  MaxIterationCount=20  MinIterationCount=15
WarmupCount=1

|                 Method |      Mean |     Error |    StdDev |    Median |       Min |       Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|------:|------:|----------:|
| Vector2.ClampBenchmark | 0.2977 ns | 0.0020 ns | 0.0018 ns | 0.2975 ns | 0.2953 ns | 0.3015 ns |     - |     - |     - |         - |
| Vector3.ClampBenchmark | 0.7103 ns | 0.0123 ns | 0.0115 ns | 0.7069 ns | 0.6934 ns | 0.7312 ns |     - |     - |     - |         - |
| Vector4.ClampBenchmark | 0.2991 ns | 0.0046 ns | 0.0043 ns | 0.2972 ns | 0.2949 ns | 0.3080 ns |     - |     - |     - |         - |
```

I just followed the workflow listed under: https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md 
The process is basically to run `.\scripts\benchmarks_ci.py --frameworks netcoreapp5.0 --filter System.Numerics.Tests.*Clamp*` and then to run again passing `--coreroot` with the path to your local build of the PR

Vector3 isn;t "as good" because it needs to additional work to load the 3 elements, where-as Vector2 and Vector4 can each be a single read/write. The codegen looks fine in other scenarios where the JIT is able to better optimize.

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
Thanks for the contribution!

**Pull request merged**